File: services/core-api/src/core_api/schemas/menu.py
"""Pydantic-схемы меню: Category, Modifier, SizeOption, MenuItem."""
# START_MODULE_CONTRACT
#   PURPOSE: Admin + public menu DTOs (categories, modifiers, sizes, items).
#            Defines the price-kopeck type and the public menu projection
#            served to non-authenticated clients.
#   SCOPE:   Base/Create/Update/Response per entity, AvailabilityPatch,
#            MenuItemModifierSet, public PublicCategory/PublicMenuItem/...
#   DEPENDS: pydantic v2, M-SHARED (CategoryType, MenuItemAvailability,
#            MenuMediaType, SizeLabel).
#   LINKS:   docs/development-plan.xml M-CORE-API, PDD §6.4 (menu state machine),
#            INV-006 (stop-list server-side), INV-010 (admin-only mutators)
#   ROLE:    TYPES
#   MAP_MODE: EXPORTS
# END_MODULE_CONTRACT
#
# START_MODULE_MAP
#   PriceKopecks            - alias type Annotated[int, Field(ge=0)]
#   InventoryQuantity       - alias type Annotated[int, Field(ge=0)]
#   MediaPath               - alias type Annotated[str, Field(...)]
#   CategoryBase            - shared category fields
#   CategoryCreate          - POST body for /admin/menu/categories
#   CategoryUpdate          - PUT body (partial) for /admin/menu/categories/{id}
#   CategoryResponse        - read projection of Category
#   ModifierBase            - shared modifier fields
#   ModifierCreate          - POST body for /admin/menu/modifiers
#   ModifierUpdate          - PUT body (partial)
#   ModifierResponse        - read projection of Modifier
#   SizeOptionBase          - shared size-option fields
#   SizeOptionCreate        - POST body for /admin/menu/sizes
#   SizeOptionUpdate        - PUT body (partial)
#   SizeOptionResponse      - read projection of SizeOption
#   MenuItemBase            - shared menu-item fields
#   MenuItemCreate          - POST body for /admin/menu/items
#   MenuItemUpdate          - PUT body (partial)
#   MenuItemResponse        - admin projection w/ size_options + modifiers
#   AvailabilityPatch       - PATCH body for stop-list toggle (extra=forbid)
#   InventoryPatch          - PATCH body for stock update (extra=forbid)
#   MenuItemModifierSet     - PUT body for setting an item's modifier list
#   PublicMenuSizeOption    - public projection of SizeOption
#   PublicMenuModifier      - public projection of Modifier
#   PublicMenuItem          - public projection of MenuItem
#   PublicCategory          - public projection of Category w/ items
#   PublicMenuResponse      - GET /api/v1/menu top-level body
# END_MODULE_MAP

from __future__ import annotations

from datetime import datetime
from typing import Annotated

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    computed_field,
    field_validator,
    model_validator,
)

from shared.enums import CategoryType, MenuItemAvailability, MenuMediaType, SizeLabel

# Общий тип для цен в копейках (>= 0)
PriceKopecks = Annotated[int, Field(ge=0)]
InventoryQuantity = Annotated[int, Field(ge=0)]
MediaPath = Annotated[str, Field(min_length=1, max_length=500)]

_MEDIA_PATH_PREFIX = "/media/menu/"
_IMAGE_EXTENSIONS = (".avif", ".jpg", ".jpeg", ".png", ".webp")
_VIDEO_EXTENSIONS = (".mp4", ".webm")


def _normalize_media_asset_path(value: object) -> object:
    if value is None:
        return None
    if not isinstance(value, str):
        return value

    path = value.strip()
    if path == "":
        return None

    lowered = path.lower()
    if not path.startswith(_MEDIA_PATH_PREFIX):
        raise ValueError("media paths must be public local paths under /media/menu/")
    if "?" in path or "#" in path:
        raise ValueError("media paths must not include query strings or fragments")
    if "://" in lowered or lowered.startswith("//"):
        raise ValueError("media paths must not be absolute external URLs")
    if any(part == ".." for part in path.split("/")):
        raise ValueError("media paths must not contain parent directory segments")
    if any(ch.isspace() for ch in path):
        raise ValueError("media paths must not contain whitespace")
    return path


def _media_type_value(media_type: MenuMediaType | str | None) -> str | None:
    if media_type is None:
        return None
    if isinstance(media_type, MenuMediaType):
        return media_type.value
    return media_type


def _has_extension(path: str, extensions: tuple[str, ...]) -> bool:
    return path.lower().endswith(extensions)


def _validate_media_combination(
    media_type: MenuMediaType | str | None,
    media_url: str | None,
    media_poster_url: str | None,
) -> None:
    media_type_value = _media_type_value(media_type)
    if media_type_value is None:
        if media_url is not None or media_poster_url is not None:
            raise ValueError("media_type is required when media URLs are set")
        return

    if media_url is None:
        raise ValueError("media_url is required when media_type is set")

    if media_type_value == MenuMediaType.VIDEO.value:
        if media_poster_url is None:
            raise ValueError("media_poster_url is required for video media")
        if not _has_extension(media_url, _VIDEO_EXTENSIONS):
            raise ValueError("video media_url must end with .mp4 or .webm")
        if not _has_extension(media_poster_url, _IMAGE_EXTENSIONS):
            raise ValueError("video media_poster_url must be an image path")
        return

    if media_type_value == MenuMediaType.IMAGE.value:
        if media_poster_url is not None:
            raise ValueError("media_poster_url is only supported for video media")
        if not _has_extension(media_url, _IMAGE_EXTENSIONS):
            raise ValueError("image media_url must be an image path")
        return

    raise ValueError("media_type must be image or video")


# ---------------------------------------------------------------------------
# Category
# ---------------------------------------------------------------------------

class CategoryBase(BaseModel):
    type: CategoryType
    name_ru: str
    name_en: str
    sort_order: int = 0
    is_visible: bool = True


class CategoryCreate(CategoryBase):
    pass


class CategoryUpdate(BaseModel):
    type: CategoryType | None = None
    name_ru: str | None = None
    name_en: str | None = None
    sort_order: int | None = None
    is_visible: bool | None = None


class CategoryResponse(CategoryBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime | None = None
    updated_at: datetime | None = None


# ---------------------------------------------------------------------------
# Modifier
# ---------------------------------------------------------------------------

class ModifierBase(BaseModel):
    name_ru: str
    name_en: str
    price: PriceKopecks = 0
    available: bool = True
    sort_order: int = 0


class ModifierCreate(ModifierBase):
    pass


class ModifierUpdate(BaseModel):
    name_ru: str | None = None
    name_en: str | None = None
    price: PriceKopecks | None = None
    available: bool | None = None
    sort_order: int | None = None


class ModifierResponse(ModifierBase):
    model_config = ConfigDict(from_attributes=True)

    id: int


# ---------------------------------------------------------------------------
# SizeOption
# ---------------------------------------------------------------------------

class SizeOptionBase(BaseModel):
    label: SizeLabel
    price: PriceKopecks = 0
    available: bool = True


class SizeOptionCreate(SizeOptionBase):
    menu_item_id: int


class SizeOptionUpdate(BaseModel):
    label: SizeLabel | None = None
    price: PriceKopecks | None = None
    available: bool | None = None


class SizeOptionResponse(SizeOptionBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    menu_item_id: int


# ---------------------------------------------------------------------------
# MenuItem  (ссылается на SizeOptionResponse и ModifierResponse — нет forward refs)
# ---------------------------------------------------------------------------

class MenuItemBase(BaseModel):
    category_id: int
    name_ru: str
    name_en: str
    description_ru: str | None = None
    description_en: str | None = None
    base_price: PriceKopecks = 0
    image_url: str | None = None
    media_type: MenuMediaType | None = None
    media_url: MediaPath | None = None
    media_poster_url: MediaPath | None = None
    inventory_quantity: InventoryQuantity | None = None
    available: bool = True
    archived: bool = False
    sort_order: int = 0

    @field_validator("media_url", "media_poster_url", mode="before")
    @classmethod
    def _normalize_media_paths(cls, value: object) -> object:
        return _normalize_media_asset_path(value)

    @model_validator(mode="after")
    def _validate_media(self) -> MenuItemBase:
        _validate_media_combination(
            self.media_type,
            self.media_url,
            self.media_poster_url,
        )
        return self


class MenuItemCreate(MenuItemBase):
    pass


class MenuItemUpdate(BaseModel):
    category_id: int | None = None
    name_ru: str | None = None
    name_en: str | None = None
    description_ru: str | None = None
    description_en: str | None = None
    base_price: PriceKopecks | None = None
    image_url: str | None = None
    media_type: MenuMediaType | None = None
    media_url: MediaPath | None = None
    media_poster_url: MediaPath | None = None
    inventory_quantity: InventoryQuantity | None = None
    available: bool | None = None
    archived: bool | None = None
    sort_order: int | None = None

    @field_validator("media_url", "media_poster_url", mode="before")
    @classmethod
    def _normalize_media_paths(cls, value: object) -> object:
        return _normalize_media_asset_path(value)


class MenuItemResponse(MenuItemBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime | None = None
    updated_at: datetime | None = None
    size_options: list[SizeOptionResponse] = []
    modifiers: list[ModifierResponse] = []

    @computed_field  # type: ignore[prop-decorator]
    @property
    def availability(self) -> MenuItemAvailability:
        if self.archived:
            return MenuItemAvailability.ARCHIVED
        if not self.available:
            return MenuItemAvailability.STOP_LIST
        return MenuItemAvailability.AVAILABLE


# ---------------------------------------------------------------------------
# AvailabilityPatch  (stop-list toggle — единственное поле)
# ---------------------------------------------------------------------------

class AvailabilityPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    available: bool


class InventoryPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    inventory_quantity: InventoryQuantity | None


class MenuItemModifierSet(BaseModel):
    model_config = ConfigDict(extra="forbid")

    modifier_ids: list[int]


# ---------------------------------------------------------------------------
# Публичный API меню (клиентская сторона)
# ---------------------------------------------------------------------------

class PublicMenuSizeOption(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    label: SizeLabel
    price: PriceKopecks
    available: bool


class PublicMenuModifier(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    name_ru: str
    name_en: str
    price: PriceKopecks
    available: bool


class PublicMenuItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    category_id: int
    name: str
    name_ru: str
    name_en: str
    description: str | None = None
    description_ru: str | None = None
    description_en: str | None = None
    base_price: PriceKopecks
    image_url: str | None = None
    media_type: MenuMediaType | None = None
    media_url: str | None = None
    media_poster_url: str | None = None
    inventory_quantity: InventoryQuantity | None = None
    available: bool
    sort_order: int
    size_options: list[PublicMenuSizeOption] = []
    modifiers: list[PublicMenuModifier] = []


class PublicCategory(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    type: CategoryType
    name: str
    name_ru: str
    name_en: str
    sort_order: int
    items: list[PublicMenuItem] = []


class PublicMenuResponse(BaseModel):
    categories: list[PublicCategory] = []

File: web/admin/src/api/menu.ts
// Типы зеркалят Pydantic-схемы из services/core-api/core_api/schemas/menu.py
// При появлении OpenAPI-кодогенерации этот файл будет заменён целиком.

import { authenticatedFetch, ApiError } from './client';

// START_MODULE_CONTRACT
//   PURPOSE: Typed client for admin menu management — categories, items,
//            sizes, modifiers — plus per-item availability toggles used by
//            barista stop-list UX.
//   SCOPE:   Wraps /api/v1/admin/menu/*; mirrors core-api Pydantic schemas in
//            services/core-api/core_api/schemas/menu.py.
//   DEPENDS: ./client (authenticatedFetch, ApiError).
//   LINKS:   docs/development-plan.xml M-WEB-ADMIN, PDD §6.4 menu CRUD,
//            INV-002 (admin scope enforced server-side; barista may only toggle
//            stop-list, full CRUD restricted at server).
//   ROLE:    RUNTIME
//   MAP_MODE: EXPORTS
// END_MODULE_CONTRACT
//
// START_MODULE_MAP
//   ApiError                 - re-export from ./client
//   Availability             - 'available' | 'stop_list' | 'archived'
//   CategoryType             - 'drink' | 'food' | 'merch' | 'modifier'
//   SizeLabel                - 'S' | 'M' | 'L'
//   CategoryResponse/...     - response/create/update DTOs (mirror Pydantic)
//   SizeOptionResponse/...   - size DTOs
//   ModifierResponse/...     - modifier DTOs
//   MenuItemResponse/...     - menu item DTOs
//   AvailabilityPatch        - PATCH body for stop-list toggle
//   listCategories           - GET /api/v1/admin/menu/categories
//   createCategory           - POST category (admin only server-side)
//   updateCategory           - PUT category (admin only)
//   deleteCategory           - DELETE category (admin only)
//   listItems                - GET /api/v1/admin/menu/items?category_id=
//   getItem                  - GET single menu item
//   createItem               - POST menu item (admin only)
//   updateItem               - PUT menu item (admin only)
//   deleteItem               - DELETE menu item (admin only)
//   setItemAvailability      - PATCH availability (admin OR barista — stop-list)
//   setItemModifiers         - PUT item-modifier link set (admin only)
//   listModifiers            - GET modifiers
//   createModifier           - POST modifier (admin only)
//   updateModifier           - PUT modifier (admin only)
//   deleteModifier           - DELETE modifier (admin only)
//   setModifierAvailability  - PATCH modifier availability (admin OR barista)
//   createSize/updateSize/deleteSize - size CRUD (admin only)
// END_MODULE_MAP

export { ApiError };

// ── Enums ────────────────────────────────────────────────────────────────────

export type Availability = 'available' | 'stop_list' | 'archived';

export type CategoryType = 'drink' | 'food' | 'merch' | 'modifier';

export type SizeLabel = 'S' | 'M' | 'L';

// ── Category ─────────────────────────────────────────────────────────────────

export interface CategoryResponse {
  id: number;
  type: CategoryType;
  name_ru: string;
  name_en: string;
  sort_order: number;
  is_visible: boolean;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface CategoryCreate {
  type: CategoryType;
  name_ru: string;
  name_en: string;
  sort_order: number;
  is_visible: boolean;
}

export interface CategoryUpdate {
  type?: CategoryType;
  name_ru?: string;
  name_en?: string;
  sort_order?: number;
  is_visible?: boolean;
}

// ── SizeOption ────────────────────────────────────────────────────────────────

export interface SizeOptionResponse {
  id: number;
  menu_item_id: number;
  label: SizeLabel;
  price: number;
  available: boolean;
}

export interface SizeOptionCreate {
  menu_item_id: number;
  label: SizeLabel;
  price: number;
  available?: boolean;
}

export interface SizeOptionUpdate {
  label?: SizeLabel;
  price?: number;
  available?: boolean;
}

// ── Modifier ──────────────────────────────────────────────────────────────────

export interface ModifierResponse {
  id: number;
  name_ru: string;
  name_en: string;
  price: number;
  available: boolean;
  sort_order: number;
}

export interface ModifierCreate {
  name_ru: string;
  name_en: string;
  price: number;
  available?: boolean;
  sort_order?: number;
}

export interface ModifierUpdate {
  name_ru?: string;
  name_en?: string;
  price?: number;
  available?: boolean;
  sort_order?: number;
}

// ── MenuItem ──────────────────────────────────────────────────────────────────

export interface MenuItemResponse {
  id: number;
  category_id: number;
  name_ru: string;
  name_en: string;
  description_ru: string | null;
  description_en: string | null;
  base_price: number;
  image_url: string | null;
  available: boolean;
  archived: boolean;
  availability: Availability;
  sort_order: number;
  size_options: SizeOptionResponse[];
  modifiers: ModifierResponse[];
  created_at?: string | null;
  updated_at?: string | null;
}

export interface MenuItemCreate {
  category_id: number;
  name_ru: string;
  name_en: string;
  description_ru?: string | null;
  description_en?: string | null;
  base_price: number;
  image_url?: string | null;
  available?: boolean;
  archived?: boolean;
  sort_order?: number;
}

export interface MenuItemUpdate {
  category_id?: number;
  name_ru?: string;
  name_en?: string;
  description_ru?: string | null;
  description_en?: string | null;
  base_price?: number;
  image_url?: string | null;
  available?: boolean;
  archived?: boolean;
  sort_order?: number;
}

// ── Availability PATCH ────────────────────────────────────────────────────────

export interface AvailabilityPatch {
  available: boolean;
}

// ── helpers ───────────────────────────────────────────────────────────────────

async function json<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await authenticatedFetch(path, init);
  if (res.status === 204) return undefined as unknown as T;
  return res.json() as Promise<T>;
}

function post<T>(path: string, body: unknown): Promise<T> {
  return json<T>(path, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
}

function put<T>(path: string, body: unknown): Promise<T> {
  return json<T>(path, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
}

function patch<T>(path: string, body: unknown): Promise<T> {
  return json<T>(path, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
}

function del(path: string): Promise<void> {
  return json<void>(path, { method: 'DELETE' });
}

// ── Categories ────────────────────────────────────────────────────────────────

// START_CONTRACT: listCategories
//   PURPOSE: Fetch all menu categories for the admin/barista UI.
//   INPUTS:  none
//   OUTPUTS: Promise<CategoryResponse[]>
//   SIDE_EFFECTS: GET; throws ApiError on non-2xx.
//   LINKS:   INV-002.
// END_CONTRACT: listCategories
export const listCategories = (): Promise<CategoryResponse[]> =>
  json('/api/v1/admin/menu/categories');

// START_CONTRACT: createCategory
//   PURPOSE: Create a new menu category (admin only — server enforces).
//   INPUTS:  body: CategoryCreate
//   OUTPUTS: Promise<CategoryResponse>
//   SIDE_EFFECTS: POST; 403 if barista tries to call.
//   LINKS:   INV-002.
// END_CONTRACT: createCategory
export const createCategory = (body: CategoryCreate): Promise<CategoryResponse> =>
  post('/api/v1/admin/menu/categories', body);

// START_CONTRACT: updateCategory
//   PURPOSE: Update a category by id (admin only).
//   INPUTS:  id: number, body: CategoryUpdate (partial)
//   OUTPUTS: Promise<CategoryResponse>
//   SIDE_EFFECTS: PUT; 403 if not admin.
//   LINKS:   INV-002.
// END_CONTRACT: updateCategory
export const updateCategory = (
  id: number,
  body: CategoryUpdate,
): Promise<CategoryResponse> => put(`/api/v1/admin/menu/categories/${id}`, body);

// START_CONTRACT: deleteCategory
//   PURPOSE: Delete a category (admin only). 409 if it still has items.
//   INPUTS:  id: number
//   OUTPUTS: Promise<void>
//   SIDE_EFFECTS: DELETE; 409 on referential conflict.
//   LINKS:   INV-002.
// END_CONTRACT: deleteCategory
export const deleteCategory = (id: number): Promise<void> =>
  del(`/api/v1/admin/menu/categories/${id}`);

// ── Menu items ────────────────────────────────────────────────────────────────

// START_CONTRACT: listItems
//   PURPOSE: List menu items, optionally filtered by category.
//   INPUTS:  params?: { categoryId?: number }
//   OUTPUTS: Promise<MenuItemResponse[]>
//   SIDE_EFFECTS: GET.
//   LINKS:   INV-002.
// END_CONTRACT: listItems
export const listItems = (params?: {
  categoryId?: number;
}): Promise<MenuItemResponse[]> => {
  const qs = params?.categoryId != null ? `?category_id=${params.categoryId}` : '';
  return json(`/api/v1/admin/menu/items${qs}`);
};

// START_CONTRACT: getItem
//   PURPOSE: Fetch a single menu item.
//   INPUTS:  id: number
//   OUTPUTS: Promise<MenuItemResponse>
//   SIDE_EFFECTS: GET.
//   LINKS:   INV-002.
// END_CONTRACT: getItem
export const getItem = (id: number): Promise<MenuItemResponse> =>
  json(`/api/v1/admin/menu/items/${id}`);

// START_CONTRACT: createItem
//   PURPOSE: Create a new menu item (admin only).
//   INPUTS:  body: MenuItemCreate
//   OUTPUTS: Promise<MenuItemResponse>
//   SIDE_EFFECTS: POST; 403 if barista calls; 422 on validation.
//   LINKS:   INV-002.
// END_CONTRACT: createItem
export const createItem = (body: MenuItemCreate): Promise<MenuItemResponse> =>
  post('/api/v1/admin/menu/items', body);

// START_CONTRACT: updateItem
//   PURPOSE: Update a menu item (admin only).
//   INPUTS:  id: number, body: MenuItemUpdate
//   OUTPUTS: Promise<MenuItemResponse>
//   SIDE_EFFECTS: PUT; 403/422.
//   LINKS:   INV-002.
// END_CONTRACT: updateItem
export const updateItem = (
  id: number,
  body: MenuItemUpdate,
): Promise<MenuItemResponse> => put(`/api/v1/admin/menu/items/${id}`, body);

// START_CONTRACT: deleteItem
//   PURPOSE: Delete a menu item (admin only).
//   INPUTS:  id: number
//   OUTPUTS: Promise<void>
//   SIDE_EFFECTS: DELETE.
//   LINKS:   INV-002.
// END_CONTRACT: deleteItem
export const deleteItem = (id: number): Promise<void> =>
  del(`/api/v1/admin/menu/items/${id}`);

// START_CONTRACT: setItemAvailability
//   PURPOSE: Toggle stop-list flag for a menu item — accessible to admin AND
//            barista (the only menu mutation barista is allowed to perform).
//   INPUTS:  id: number, available: boolean
//   OUTPUTS: Promise<MenuItemResponse> — fresh availability state from server.
//   SIDE_EFFECTS: PATCH.
//   LINKS:   INV-002 (server allows admin+barista here, denies courier),
//            INV-010 (role isolation — barista cannot edit anything else).
// END_CONTRACT: setItemAvailability
export const setItemAvailability = (
  id: number,
  available: boolean,
): Promise<MenuItemResponse> =>
  patch(`/api/v1/admin/menu/items/${id}/availability`, { available });

// START_CONTRACT: setItemModifiers
//   PURPOSE: Replace the modifier set linked to a menu item (admin only).
//   INPUTS:  id: number, modifier_ids: number[]
//   OUTPUTS: Promise<MenuItemResponse>
//   SIDE_EFFECTS: PUT; 403 if barista calls.
//   LINKS:   INV-002.
// END_CONTRACT: setItemModifiers
export const setItemModifiers = (
  id: number,
  modifier_ids: number[],
): Promise<MenuItemResponse> =>
  put(`/api/v1/admin/menu/items/${id}/modifiers`, { modifier_ids });

// ── Modifiers ─────────────────────────────────────────────────────────────────

// START_CONTRACT: listModifiers
//   PURPOSE: Fetch all modifiers for picker UIs and stop-list management.
//   INPUTS:  none
//   OUTPUTS: Promise<ModifierResponse[]>
//   SIDE_EFFECTS: GET.
//   LINKS:   INV-002.
// END_CONTRACT: listModifiers
export const listModifiers = (): Promise<ModifierResponse[]> =>
  json('/api/v1/admin/menu/modifiers');

// START_CONTRACT: createModifier
//   PURPOSE: Create a new modifier (admin only).
//   INPUTS:  body: ModifierCreate
//   OUTPUTS: Promise<ModifierResponse>
//   SIDE_EFFECTS: POST; 403/422.
//   LINKS:   INV-002.
// END_CONTRACT: createModifier
export const createModifier = (body: ModifierCreate): Promise<ModifierResponse> =>
  post('/api/v1/admin/menu/modifiers', body);

// START_CONTRACT: updateModifier
//   PURPOSE: Update a modifier (admin only).
//   INPUTS:  id: number, body: ModifierUpdate
//   OUTPUTS: Promise<ModifierResponse>
//   SIDE_EFFECTS: PUT.
//   LINKS:   INV-002.
// END_CONTRACT: updateModifier
export const updateModifier = (
  id: number,
  body: ModifierUpdate,
): Promise<ModifierResponse> => put(`/api/v1/admin/menu/modifiers/${id}`, body);

// START_CONTRACT: deleteModifier
//   PURPOSE: Delete a modifier (admin only).
//   INPUTS:  id: number
//   OUTPUTS: Promise<void>
//   SIDE_EFFECTS: DELETE.
//   LINKS:   INV-002.
// END_CONTRACT: deleteModifier
export const deleteModifier = (id: number): Promise<void> =>
  del(`/api/v1/admin/menu/modifiers/${id}`);

// START_CONTRACT: setModifierAvailability
//   PURPOSE: Toggle stop-list flag on a modifier — admin OR barista (parallel
//            to setItemAvailability).
//   INPUTS:  id: number, available: boolean
//   OUTPUTS: Promise<ModifierResponse>
//   SIDE_EFFECTS: PATCH.
//   LINKS:   INV-002, INV-010.
// END_CONTRACT: setModifierAvailability
export const setModifierAvailability = (
  id: number,
  available: boolean,
): Promise<ModifierResponse> =>
  patch(`/api/v1/admin/menu/modifiers/${id}/availability`, { available });

// ── Sizes ─────────────────────────────────────────────────────────────────────

// START_CONTRACT: createSize
//   PURPOSE: Add a size option (S/M/L) to a menu item (admin only).
//   INPUTS:  body: SizeOptionCreate
//   OUTPUTS: Promise<SizeOptionResponse>
//   SIDE_EFFECTS: POST; 409 on duplicate label.
//   LINKS:   INV-002.
// END_CONTRACT: createSize
export const createSize = (body: SizeOptionCreate): Promise<SizeOptionResponse> =>
  post('/api/v1/admin/menu/sizes', body);

// START_CONTRACT: updateSize
//   PURPOSE: Update a size option (admin only).
//   INPUTS:  id: number, body: SizeOptionUpdate
//   OUTPUTS: Promise<SizeOptionResponse>
//   SIDE_EFFECTS: PUT.
//   LINKS:   INV-002.
// END_CONTRACT: updateSize
export const updateSize = (
  id: number,
  body: SizeOptionUpdate,
): Promise<SizeOptionResponse> => put(`/api/v1/admin/menu/sizes/${id}`, body);

// START_CONTRACT: deleteSize
//   PURPOSE: Delete a size option (admin only).
//   INPUTS:  id: number
//   OUTPUTS: Promise<void>
//   SIDE_EFFECTS: DELETE.
//   LINKS:   INV-002.
// END_CONTRACT: deleteSize
export const deleteSize = (id: number): Promise<void> =>
  del(`/api/v1/admin/menu/sizes/${id}`);
